### PR TITLE
[WFLY-14655] Add proper ordering of ServiceMBeanSupport postRegister calls

### PR DESCRIPTION
**Pull request by bstansberry**
https://issues.redhat.com/browse/WFLY-14655

### CODE REVIEW

**Comment by bstansberry** [comment]
I just updated this PR to replace the cut-and-pasted class javadoc in the test class I added with the correct content. There was no functional change.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/LifecycleListener.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.List;
+
+import org.jboss.system.ServiceMBeanSupport;
+
+public class LifecycleListener extends ServiceMBeanSupport implements LifecycleListenerMBean {
+
+    private final List<String> creates = new ArrayList<>();
+    private final List<String> starts = new ArrayList<>();
+    private final List<String> stops = new ArrayList<>();
+    private final List<String> destroys = new ArrayList<>();
```
Comment (pferraro) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/LifecycleListener.java:31]: Are these collections vulnerable to ConcurrentModificationExceptions?

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/ServiceMBeanSupportOrderingTestCase.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.container.test.api.RunAsClient;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.as.arquillian.api.ContainerResource;
+import org.jboss.as.arquillian.container.ManagementClient;
+import org.jboss.as.test.integration.common.DefaultConfiguration;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.system.ServiceMBean;
```
Comment (pferraro) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/ServiceMBeanSupportOrderingTestCase.java:51]: Unused import is failing checkstyle.

**Review by pferraro — CHANGES_REQUESTED** [review]

**Comment by pferraro** [comment]
Other than the above, the fix itself looks good.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/LifecycleListener.java (hunk 1 above)
Comment (bstansberry) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/LifecycleListener.java:31]: Good point.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/LifecycleListener.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    private final List<String> stops = Collections.synchronizedList(new ArrayList<>());
+    private final List<String> destroys = Collections.synchronizedList(new ArrayList<>());
+
+    @Override
+    public void mbeanCreated(String id) {
+        creates.add(id);
+    }
+
+    @Override
+    public synchronized void mbeanStarted(String id) {
```
Comment (pferraro) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/LifecycleListener.java:40]: Nit: This synchronized modifier is no longer necessary.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/LifecycleListener.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        starts.add(id);
+    }
+
+    @Override
+    public synchronized void mbeanStopped(String id) {
```
Comment (pferraro) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/sar/order/LifecycleListener.java:45]: Nit: This synchronized modifier is no longer necessary.

**Review by pferraro — APPROVED** [review]
Just a couple of nits, which don't need to block merging this.

**Pull request merged**